### ISSUE TITLE
[question] when conan install is called inside a dockerfile, cmake cannot resolve dependencies

### ISSUE THREAD

**Issue by sdeadric** [conan-io/conan#14438]
### What is your question?

I have a project where I call conan install inside a dockerfile so that I have the dependencies needed to compile with cmake inside the docker container. However, after install is called when starting the container, the call to cmake results in `Library XXX not found in package`. If I then delete the build folder and call conan install again, cmake can then find the dependencies. What might cause this?

conan install called in dockerfile
`conan install . --build="*" -s compiler.cppstd=gnu20`

conan profile inside container
```
[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.cppstd=gnu14
compiler.libcxx=libstdc++
compiler.version=10
os=Linux
```

conan profile outside container
```
[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.cppstd=gnu17
compiler.libcxx=libstdc++11
compiler.version=11
os=Linux
```

conanfile.py
```
import os
from conan import ConanFile
from conan.tools.cmake import CMakeToolchain, CMake, cmake_layout
from conan.tools.files import get
from conan.tools.build import check_min_cppstd

class ProjectRecipe(ConanFile):
    name = 'data-fusion-engine'

    settings = 'os', 'compiler', 'build_type', 'arch'
    options = {
        'shared': [True, False],
        'fPIC': [True, False]
    }

    default_options = {
        'shared': False,
        'fPIC': True
    }

    generators = 'CMakeDeps'

    exports_sources = 'CMakeLists.txt', 'src/*'

    def validate(self):
        check_min_cppstd(self, '20')

    def requirements(self):
        self.requires('cspice/0067')

    def build_requirements(self):
        self.tool_requires('cmake/3.27.0')

    def config_options(self):
        if self.settings.os == 'Windows':
            del self.options.fPIC

    def configure(self):
        if self.options.shared:
            self.options.rm_safe('fPIC')

    def layout(self):
        cmake_layout(self)

    def generate(self):
        tc = CMakeToolchain(self)
        tc.generate()

    def build(self):
        cmake = CMake(self)
        cmake.configure()
        cmake.build()
        if not self.conf.get('tools.build:skip_test', default=False):
            import os

    def package(self):
        cmake = CMake(self)
        cmake.install(self)

    def package_info(self):
        self.cpp_info.libs = ['project']
        self.cpp_info.includedirs = ['include']
        self.cpp_info.libdirs = ['lib']
```

I have tried:
- implementing a conan.lock
- deleting the build folder inside the dockerfile before calling conan install
- cleaning the conan cache inside the dockerfile
- matching the conan profiles in and outside the container via -s flags in the conan install command

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @sdeadric 

Thanks for your question.

There are some details missing:
- the Conan version (2.0 recommended)
- details about the docker image, the command to launch it, if you are sharing/mounting some volume, specially if it is the Conan cache and if it would be shared by the host or something like that.
- If you can share the docker image, please do
- What is failing to compile the recipe that you are sharing above, or the dependency ``cspice/0067``
- If your local recipe is the one failing to build, I am missing commands, because ``conan install`` is not building the local recipe.
- Please share the full output of the failing command.

**Comment by sdeadric** [comment]
- Conan version is 2.0
- The container is built on a RHEL7 image
- The command vscode uses to launch the container is `docker-compose --project-name project -f /home/user/project/docker-compose.yml -f /home/user/project/.devcontainer/docker-compose.yml -f /tmp/devcontainercli-user/docker-compose/docker-compose.devcontainer.build-1691453513080.yml build`
- mounted volumes include
```
    volumes:
      - ./kernels:/project/kernels
      - ../test-cases/test_folders:/project/src/testJSON
      - clangd-cache:/project/.cache
      - cmake-cache:/project/build
```
- The error occurs when vscode attempts to build using the command 
```
/usr/local/bin/cmake -DCMAKE_POLICY_DEFAULT_CMP0091=NEW -DCMAKE_BUILD_TYPE=Release -DCMAKE_TOOLCHAIN_FILE=/project/build/Release/generators/conan_toolchain.cmake -S/project -B/project/build/Release -G "Unix Makefiles"
```
- The full output is
```
[main] Configuring project: project 
[proc] Executing command: /usr/local/bin/cmake -DCMAKE_POLICY_DEFAULT_CMP0091=NEW -DCMAKE_BUILD_TYPE=Release -DCMAKE_TOOLCHAIN_FILE=/project/build/Release/generators/conan_toolchain.cmake -S/project -B/project/build/Release -G "Unix Makefiles"
[cmake] -- /usr/local/include
[cmake] -- Using Conan toolchain: /project/build/Release/generators/conan_toolchain.cmake
[cmake] -- Conan toolchain: Setting CMAKE_POSITION_INDEPENDENT_CODE=ON (options.fPIC)
[cmake] -- Conan toolchain: C++ Standard 20 with extensions ON
[cmake] -- Conan toolchain: Setting BUILD_SHARED_LIBS = OFF
[cmake] -- Conan: Component target declared 'Eigen3::Eigen'
[cmake] -- Conan: Target declared 'cspice::cspice'
[cmake] CMake Error at build/Release/generators/cmakedeps_macros.cmake:67 (message):
[cmake]   Library 'cspice' not found in package.  If 'cspice' is a system library,
[cmake]   declare it with 'cpp_info.system_libs' property
[cmake] Call Stack (most recent call first):
[cmake]   build/Release/generators/cspice-Target-release.cmake:24 (conan_package_library_targets)
[cmake]   build/Release/generators/cspiceTargets.cmake:26 (include)
[cmake]   build/Release/generators/cspice-config.cmake:16 (include)
[cmake]   CMakeLists.txt:45 (find_package)
[cmake] 
[cmake] 
[cmake] -- Configuring incomplete, errors occurred!
[cmake] See also "/project/build/Release/CMakeFiles/CMakeOutput.log".
[cmake] See also "/project/build/Release/CMakeFiles/CMakeError.log".
[proc] The command: /usr/local/bin/cmake -DCMAKE_POLICY_DEFAULT_CMP0091=NEW -DCMAKE_BUILD_TYPE=Release -DCMAKE_TOOLCHAIN_FILE=/project/build/Release/generators/conan_toolchain.cmake -S/project -B/project/build/Release -G "Unix Makefiles" exited with code: 1
```

**Comment by memsharded** [comment]
Thanks for the details @sdeadric 

I think it is possible that the mounted volumes, maybe the ``cmake-cache:/project/build`` could be highly related, we have seen similar things in the past. I'd suggest trying:

- Making sure that those folders are completely empty before starting
- Trying to run it with the build folder not mounted, just private inside the container, and see if it works.

**Comment by sdeadric** [comment]
When I unmount `cmake-cache:/project/build`, I now get this error output
```
[main] Building folder: project 
[main] Configuring project: project  
[proc] Executing command: /usr/local/bin/cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=/project/out/install/conan-release -S/project -B/project/out/build/conan-release -G "Unix Makefiles"
[cmake] -- /usr/local/include
[cmake] -- The CXX compiler identification is GNU 10.2.1
[cmake] -- The C compiler identification is GNU 10.2.1
[cmake] -- Detecting CXX compiler ABI info
[cmake] -- Detecting CXX compiler ABI info - done
[cmake] -- Check for working CXX compiler: /opt/rh/devtoolset-10/root/usr/bin/c++ - skipped
[cmake] -- Detecting CXX compile features
[cmake] -- Detecting CXX compile features - done
[cmake] -- Detecting C compiler ABI info
[cmake] -- Detecting C compiler ABI info - done
[cmake] -- Check for working C compiler: /opt/rh/devtoolset-10/root/usr/bin/cc - skipped
[cmake] -- Detecting C compile features
[cmake] -- Detecting C compile features - done
[cmake] -- Looking for pthread.h
[cmake] -- Looking for pthread.h - found
[cmake] -- Performing Test CMAKE_HAVE_LIBC_PTHREAD
[cmake] -- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
[cmake] -- Looking for pthread_create in pthreads
[cmake] -- Looking for pthread_create in pthreads - not found
[cmake] -- Looking for pthread_create in pthread
[cmake] -- Looking for pthread_create in pthread - found
[cmake] -- Found Threads: TRUE  
[cmake] -- Found OpenSSL: /usr/lib64/libcrypto.so (found version "1.0.2k")  
[cmake] CMake Error at CMakeLists.txt:45 (find_package):
[cmake]   By not providing "Findcspice.cmake" in CMAKE_MODULE_PATH this project has
[cmake]   asked CMake to find a package configuration file provided by "cspice", but
[cmake]   CMake did not find one.
[cmake] 
[cmake]   Could not find a package configuration file provided by "cspice" with any
[cmake]   of the following names:
[cmake] 
[cmake]     cspiceConfig.cmake
[cmake]     cspice-config.cmake
[cmake] 
[cmake]   Add the installation prefix of "cspice" to CMAKE_PREFIX_PATH or set
[cmake]   "cspice_DIR" to a directory containing one of the above files.  If "cspice"
[cmake]   provides a separate development package or SDK, be sure it has been
[cmake]   installed.
[cmake] 
[cmake] 
[cmake] -- Configuring incomplete, errors occurred!
[cmake] See also "/project/out/build/conan-release/CMakeFiles/CMakeOutput.log".
[cmake] See also "/project/out/build/conan-release/CMakeFiles/CMakeError.log".
[proc] The command: /usr/local/bin/cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=/project/out/install/conan-release -S/project -B/project/out/build/conan-release -G "Unix Makefiles" exited with code: 1
```

**Comment by sdeadric** [comment]
I have since noticed that I do not even need to delete the build folder; if I simply run the conan install command a second time from a terminal inside the running container I am able to compile after.

**Comment by memsharded** [comment]
``usr/local/bin/cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=/project/out/install/conan-release -S/project -B/project/out/build/conan-release -G "Unix Makefiles"`` does not use the ``conan_toolchain.cmake``, that contains the paths to locate the ``xxx-config.cmake`` files created by ``CMakeDeps`` to locate the dependencies.

We might be missing something here.

**Comment by sdeadric** [comment]
I am not sure how vscode determines what flags to append to the call to cmake, this most recent time it ran:
```
/usr/local/bin/cmake -DCMAKE_POLICY_DEFAULT_CMP0091=NEW -DCMAKE_BUILD_TYPE=Release -DCMAKE_TOOLCHAIN_FILE=/project/build/Release/generators/conan_toolchain.cmake -S/project -B/project/build/Release -G "Unix Makefiles"
```

But it still gave the error from [here](https://github.com/conan-io/conan/issues/14438#issuecomment-1668741488). I also noticed the 
package names in the conan_toolchain.cmake `CMAKE_LIBRARY_PATH` and `CMAKE_INCLUDE_PATH` file do not match those actually found in `/opt/app-root/src/.conan2/p/b`. Even when I do update them to the correct names, though, I get that same error.

**Comment by memsharded** [comment]
Maybe it is something in your CMakeLists?
At this point I think we need something to reproduce. The best would be:

- Put the relevant code to reproduce in a GitHub repo. The consumer ``conanfile.py``, consumer ``CMakeLists.txt``
- Include a small script to reproduce that calls the exact commands that fire this
- Also, I haven't asked, but just in case? Does this work in a regular Linux computer, no docker involved?

**Comment by AbrilRBS** [comment]
Closing as stale as we didn't hear back from you. If anyone finds themselves in a similar situation, do feel free to create a new issue, we're always happy to help :)
